src/java_tools/buildjar/java/com/google/devtools/build/java/turbine/BUILD

```diff
@@ -16,6 +18,31 @@ java_binary(
     ],
 )
 
+java_binary(
+    name = "turbine_direct_binary",
+    main_class = _TURBINE_MAIN_CLASS,
+    runtime_deps = [":turbine_deps"],
+)
+
+native_image(
+    name = "turbine_direct_graal",
+    executable_name = select({
+        "@bazel_tools//src/conditions:windows": "%target%.exe",
+        "//conditions:default": "%target%",
+    }),
+    extra_args = select({
+        "@platforms//os:linux": [
+            # Statically link libz but not glibc.
+            "-H:+StaticExecutableWithDynamicLibC",
+        ],
```

Comment:
These are some other flags we're using at Google. I don't know if these matter for Bazel, so this is more an FYI in case they end up mattering for Bazel, not a request to add all of them.

* `-H:-UseContainerSupport` - fixes an NPE running in containers (https://github.com/oracle/graal/issues/4757)
* `-H:+ReportExceptionStackTraces` - "Show exception stack traces for exceptions during image building" - might make failures easier to debug
* `--no-fallback` - build a native image or fail, never fall back to using a full VM
* `-R:MaxHeapSize=3g` / `-R:MinHeapSize=3g` - this significantly improved performance for us in ~2022, I haven't revisited tuning it. It might be worth tuning heap sizes and seeing how that affects your benchmarking at some point.

---

Additionally, we're currently linking a handful of annotation processors that are widely used at Google into the native image to avoid having to use a VM for those. That requires a bit of set up of `-H:DynamicProxyConfigurationFiles=` and `-H:ReflectionConfigurationFiles=`.

It probably doesn't make sense to start baking specific annotation processors into the default Bazel toolchains, but it might be something that is beneficial for Bazel deployments that make extensive use of annotation processing, and are willing to build custom native images.

Reply:
@cushon I wonder if a rule would be appropriate in `rules_graalvm` to produce pre-built annotation processor binaries?

Reply:
@cushon Are there any public processors that are very common internally (e.g. for Android) and that would make sense to support out-of-the-box? I personally don't have a need for any though.

Reply:
The public ones we're building in are https://github.com/google/dagger and https://github.com/google/auto/tree/main/factory, Dagger is widely used for Android at Google, I'm not sure what usage looks like outside Google. I'm also not sure how sense out-of-the-box support makes, it might be too surprising to get a specific version of those processors bundled with Bazel.

Reply:
If we really wanted to, we could even use a transition to a bootstrap Java toolchain with a Turbine JAR to automatically build a Graal native image and assemble it into a `java_toolchain` for a given set of `java_plugin` targets. But I wouldn't go there without a strong indication that anybody needs this. For now, I would just add some docs on how to build Turbine in a follow-up change.

@cushon I am a bit worried that `-R:MinHeapSize=3g` would cause Bazel to overschedule these actions when running locally. In https://github.com/bazelbuild/bazel/blob/a239ea84832f18ee8706682145e9595e71b39680/src/main/java/com/google/devtools/build/lib/rules/java/JavaCompileAction.java#L102, all compilation actions are assumed to consume ~750 MB only. Is that something you have seen cause issues?

Reply:
> For now, I would just add some docs on how to build Turbine in a follow-up change.

+1

> @cushon I am a bit worried that `-R:MinHeapSize=3g` would cause Bazel to overschedule these actions when running locally. In
> 
> https://github.com/bazelbuild/bazel/blob/a239ea84832f18ee8706682145e9595e71b39680/src/main/java/com/google/devtools/build/lib/rules/java/JavaCompileAction.java#L102
> 
> , all compilation actions are assumed to consume ~750 MB only. Is that something you have seen cause issues?

I haven't seen that cause issues, but Blaze is generally not running these actions locally. I suspect the tuning I did was helpful mainly for large compilations with a lot of generated code.

It sounds like you're seeing a nice speedup with the defaults. I think it'd be reasonable to stick with the defaults for now and consider tuning that later if there's evidence it would help for Bazel.